### PR TITLE
[LLM Runtime] GGUF Enhencement

### PR DESCRIPTION
**Pull request by Zhenzhong1**
## Type of Change

This PR is the GGUF Enhencement for https://github.com/intel/neural-speed/pull/40.

New feature add

- Support loading & inference official GGUF models from HF.
- Support loading & inference official GGUF models from llama.cpp.

## Description: 

Verified GGUF models from HF:

- [TheBloke/Llama-2-7B-Chat-GGUF](https://huggingface.co/TheBloke/Llama-2-7B-Chat-GGUF) : q4_0 / q5_0 / q8_0
- [TheBloke/Mistral-7B-v0.1-GGUF](https://huggingface.co/TheBloke/Mistral-7B-v0.1-GGUF) : q4_0 / q5_0 / q8_0
- [TheBloke/CodeLlama-7B-GGUF](https://huggingface.co/TheBloke/CodeLlama-7B-GGUF) : q4_0 / q5_0 / q8_0
- [TheBloke/CodeLlama-13B-GGUF](https://huggingface.co/TheBloke/CodeLlama-13B-GGUF): q4_0 / q5_0 / q8_0

Verified GGUF models from Llama.cpp:

- [meta-llama/Llama-2-7b-chat-hf](https://huggingface.co/meta-llama/Llama-2-7b-chat-hf) : f32 / f16 / BTLA
- [tiiuae/falcon-7b](https://huggingface.co/tiiuae/falcon-7b/tree/main) : f32 / f16 / BTLA
- [tiiuae/falcon-40b](https://huggingface.co/tiiuae/falcon-40b) : f32 / f16 / BTLA
- [mpt-7b](https://huggingface.co/mosaicml/mpt-7b) : f32 / f16 / BTLA
- [mpt-30b](https://huggingface.co/mosaicml/mpt-30b) : f32 / f16 / BTLA
- [bloomz-7b1](https://huggingface.co/bigscience/bloomz-7b1) : f32 / f16 / BTLA

How to create the GGUF bin file in NeuralSpeed:
```python
# Example:
# please provide the local model path as the arg, 
# which means you need to `git clone https://huggingface.co/meta-llama/Llama-2-7b-chat-hf` first.
python neural_speed/convert/convert-hf-to-gguf.py /model_path/Llama-2-7b-chat-hf/

```

How to load the GGUF bin file in NeuralSpeed:

```python
    prompt = "Once upon a time"
    tokenizer = AutoTokenizer.from_pretrained(args.model_path, trust_remote_code=True)
    inputs = tokenizer(prompt, return_tensors="pt").input_ids
    streamer = TextStreamer(tokenizer)

    model = Model()
    model.init_from_bin(args.model_name, gguf_path)
    outputs = model.generate(inputs, streamer=streamer, max_new_tokens=300, do_sample=True)

# Please check this script for more details and input parameters.
# python scripts/python_api_example_for_gguf.py --model_name falcon --model_path /home/model/falcon-7b -m /home/model/falcon-7b/ggml-model-f32.gguf

```

Note: These GGUF models from Llama.cpp can be accelerated by [NeuralSpeed BestTLA](https://github.com/intel/neural-speed/blob/c0312283f528d4a9ffebc283cd0f15a7a8eabf1a/bestla/README.md#L1).

How to accelerate GGUF by BTLA:
```python

# quantization and re-run the python_api_example_for_gguf.py
./build/bin/quant_falcon --model_file /home/model/falcon-7b/ggml-model-f32.gguf --out_file ne-falcon-q4_j.bin --weight_dtype int4 --compute_dtype int8
```

How to load the GGUF bin file in [intel-extension-for-transformers](https://github.com/intel/intel-extension-for-transformers/pull/1151):
```python
from transformers import AutoTokenizer, TextStreamer
from intel_extension_for_transformers.transformers import AutoModelForCausalLM, WeightOnlyQuantConfig

# Specify the GGUF repo on the Hugginface
model_name = "TheBloke/Llama-2-7B-Chat-GGUF"
# Download the the specific gguf model file from the above repo
model_file = "llama-2-7b-chat.Q4_0.gguf"
# make sure you are granted to access this model on the Huggingface.
tokenizer_name = "meta-llama/Llama-2-7b-chat-hf"

prompt = "Once upon a time"
tokenizer = AutoTokenizer.from_pretrained(tokenizer_name, trust_remote_code=True)
inputs = tokenizer(prompt, return_tensors="pt").input_ids
streamer = TextStreamer(tokenizer)
model = AutoModelForCausalLM.from_pretrained(model_name, model_file = model_file)
outputs = model.generate(inputs, streamer=streamer, max_new_tokens=300)
```

## Expected Behavior & Potential Risk

N/A

## How has this PR been tested?

CI tests.

## Dependency Change?

N/A

### CODE REVIEW

**Comment by Zhenzhong1** [comment]
https://inteltf-jenk.sh.intel.com/job/neural_speed_extension/28/

File: neural_speed/models/llama/llama_utils.cpp [hunk 1]
```diff
@@ -112,41 +113,78 @@ void Llama::load(model_context* ctx, model_progress_callback progress_callback,
...
-    // ffn GEMM
-    layer.ffn[0] = ml->get_tensor(layers_i + ".feed_forward.w1.weight", {n_embd, n_ff}, backend);
-    layer.ffn[1] = ml->get_tensor(layers_i + ".feed_forward.w2.weight", {n_ff, n_embd}, backend);
-    layer.ffn[2] = ml->get_tensor(layers_i + ".feed_forward.w3.weight", {n_embd, n_ff}, backend);
-
-    if (backend != NE_BACKEND_CPU) {
-      vram_total += ne_nbytes(layer.norm[0]) + ne_nbytes(layer.attn[0]) + ne_nbytes(layer.attn[1]) +
-                    ne_nbytes(layer.attn[2]) + ne_nbytes(layer.attn[3]) + ne_nbytes(layer.norm[1]) +
-                    ne_nbytes(layer.ffn[0]) + ne_nbytes(layer.ffn[1]) + ne_nbytes(layer.ffn[2]);
+  if (ml->verify_tensor("token_embd.weight")) {
```
Comment (zhenwei-intel) [inline neural_speed/models/llama/llama_utils.cpp:119]: why different tensor name here?

File: scripts/python_api_example_for_gguf.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+from pathlib import Path
+from typing import List, Optional
+from transformers import AutoTokenizer, TextStreamer
+from neural_speed import Model
+
+# Usage:
+# python python_api_example_for_bin.py \
+# --model_name falcon \
+# --model_path /home/zhenzhong/model/falcon-7b \
+# -m /home/zhenzhong/model/falcon-7b/ggml-model-f32.gguf
```
Comment (zhenwei-intel) [inline scripts/python_api_example_for_gguf.py:28]:
```suggestion
# -m /path/to/ggml-model-f32.gguf
```

File: neural_speed/models/llama/llama_utils.cpp (hunk 1 above)
Comment (Zhenzhong1) [reply neural_speed/models/llama/llama_utils.cpp:119]: becuase the convert of llama.cpp unifies weight names of different models. It's not the same as our previous weight names.

**Review by zhenwei-intel — APPROVED** [review]

**Review by luoyu-intel — APPROVED** [review]

**Review by a32543254 — APPROVED** [review]
LGTM

File: README.md [hunk 3]
```diff
@@ -473,3 +473,85 @@ Available modes:
...
+
+Verified GGUF models from HF:
+
+- [TheBloke/Llama-2-7B-Chat-GGUF](https://huggingface.co/TheBloke/Llama-2-7B-Chat-GGUF) : q4_0 / q5_0 / q8_0
+- [TheBloke/Mistral-7B-v0.1-GGUF](https://huggingface.co/TheBloke/Mistral-7B-v0.1-GGUF) : q4_0 / q5_0 / q8_0
+- [TheBloke/CodeLlama-7B-GGUF](https://huggingface.co/TheBloke/CodeLlama-7B-GGUF) : q4_0 / q5_0 / q8_0
+- [TheBloke/CodeLlama-13B-GGUF](https://huggingface.co/TheBloke/CodeLlama-13B-GGUF): q4_0 / q5_0 / q8_0
+
+
+Verified GGUF models from Llama.cpp:
```
Comment (a32543254) [inline README.md:492]: no put GGUF from hf or from llama.cpp in here.

File: README.md [hunk 4]
```diff
@@ -473,3 +473,85 @@ Available modes:
...
+# Specify the GGUF repo on the Hugginface
+model_name = "TheBloke/Llama-2-7B-Chat-GGUF"
+# Download the the specific gguf model file from the above repo
+model_file = "llama-2-7b-chat.Q4_0.gguf"
+# make sure you are granted to access this model on the Huggingface.
+tokenizer_name = "meta-llama/Llama-2-7b-chat-hf"
+
+# If you are not granted, please check this combination.
+# model_name = "TheBloke/Mistral-7B-v0.1-GGUF"
+# model_file = "mistral-7b-v0.1.Q4_0.gguf"
```
Comment (a32543254) [inline README.md:548]: choose llama or mistral

File: neural_speed/convert/convert_chatglm.py [hunk 5]
```diff
@@ -155,38 +152,44 @@ def chatglm2_convert_gguf(model, tokenizer, dir_model, fname_out, ftype, hparams
...
+    gguf_writer.add_feed_forward_length(hparams["ffn_hidden_size"])
+
     gguf_writer.add_uint32('inner_hidden_size', 0)
 
-    gguf_writer.add_int32('bos_token_id', tokenizer.bos_token_id if tokenizer.bos_token_id is not None else -1)
-    gguf_writer.add_int32('eos_token_id', tokenizer.eos_token_id if tokenizer.eos_token_id is not None else -1)
-    gguf_writer.add_int32('pad_token_id', tokenizer.pad_token_id if tokenizer.pad_token_id is not None else -1)
-    gguf_writer.add_int32('sep_token_id', tokenizer.sep_token_id if tokenizer.sep_token_id is not None else -1)
+    gguf_writer.add_bos_token_id(tokenizer.bos_token_id if tokenizer.bos_token_id is not None else 0)
+    gguf_writer.add_eos_token_id(tokenizer.eos_token_id if tokenizer.eos_token_id is not None else 0)
```
Comment (a32543254) [inline neural_speed/convert/convert_chatglm.py:190]: we cannot set bos and eos as 0 as default, beacuse some time the 0 token id mean some word

File: README.md (hunk 4 above)
Comment (Zhenzhong1) [reply README.md]: fixed

File: neural_speed/convert/convert_chatglm.py (hunk 5 above)
Comment (Zhenzhong1) [reply neural_speed/convert/convert_chatglm.py:190]: This script will not affect other models. only chatglm2.
The defaul eos_token_id is 2 and bos_token_id is 0.

File: README.md (hunk 3 above)
Comment (Zhenzhong1) [reply README.md]: fixed

**Pull request merged**